documentation/specs/src/base-ledger/replay-protection.md

```diff
@@ -0,0 +1,503 @@
+
+### Hash register
+
+The actual Wasm code and data for the transaction are encapsulated inside a struct `Tx`, which gets encrypted as an `EncryptedTx` and wrapped inside a `WrapperTx` (see the [relative](#encryption-authentication) section). This inner transaction must be protected from replay attacks because it carries the actual semantics of the state transition. Moreover, even if the wrapper transaction was protected from replay attacks, an attacker could extract the inner transaction, rewrap it, and replay it. Note that for this attack to work, the attacker will need to sign the outer transaction himself and pay gas and fees for that, but this could still cause much greater damage to the parties involved in the inner transaction.
+
+`WrapperTx` is the only type of transaction currently accepted by the ledger. It must be protected from replay attacks because, if it wasn't, a malicious user could replay the transaction as is. Even if the inner transaction implemented replay protection or, for any reason, wasn't accepted, the signer of the wrapper would still pay for gas and fees, effectively suffering economic damage.
+
+To prevent the replay of both these transactions we will rely on a set of already processed transactions' digests that will be kept in storage. These digests will be computed on the **signed** transactions. To support this, we'll need a subspace in storage headed by a `ReplayProtection` internal address:
+
+```
+/$ReplayProtectionAddress/$tx0_hash: None
+/$ReplayProtectionAddress/$tx1_hash: None
+/$ReplayProtectionAddress/$tx2_hash: None
+...
+```
+
+The hashes will form the last part of the path to allow for a fast storage lookup.
+
+The consistency of the storage subspace is of critical importance for the correct working of the replay protection mechanism. To protect it, a validity predicate will check that no changes to this subspace are applied by any wasm transaction, as those should only be available from protocol.
+
+Both in `mempool_validation` and `process_proposal` we will perform a check (together with others, see the [relative](#wrapper-checks) section) on both the digests against the storage to check that neither of the transactions has already been executed: if this doesn't hold, the `WrapperTx` will not be included into the mempool/block respectively. If both checks pass then the transaction is included in the block and executed. In the `finalize_block` function we will add the transaction's hash to storage to prevent re-executions. We will first add the hash of the wrapper transaction. After that, in the following block, we deserialize the inner transaction, check the correct order of the transactions in the block and execute the tx: if it runs out of gas then we'll avoid storing its hash to allow rewrapping and executing the transaction, otherwise we'll add the hash in storage (both in case of success or failure of the tx).
```

Comment:
Hmm I don't think this is right - if the transaction is executed and pays some fee, but runs out of gas, we should still prevent replay (the user can issue a new transaction)

otherwise the user may pay more fees than intended

Reply:
Actually, by the time we get to the execution of the inner transaction, the wrapper hash has already been saved to storage (in the previous block), so you can't replay the wrapper. You can only take the encrypted inner transaction and include it in a new wrapper

Reply:
I see. I think this is probably fine but we should put a TTL in inner transactions then, it's weird to have some signed data around for a long time which could potentially be executed at any moment

Reply:
We do that with this last spec revision. The struct `Tx` will carry an `expiration` field that will define the lifetime of the transaction. The user will be able to set this value and we'll perform a set of checks on it. Since it's implemented in struct `Tx`, this will apply to both wrappers and inners.

Reply:
OK sounds good